ipfs-cluster-service/state.go

```diff
@@ -32,38 +38,55 @@ func upgrade() error {
 }
 
 func export(w io.Writer) error {
-	stateToExport, err := restoreStateFromDisk()
+	stateToExport, _, err := restoreStateFromDisk()
 	if err != nil {
 		return err
 	}
 
 	return exportState(stateToExport, w)
 }
 
-func restoreStateFromDisk() (*mapstate.MapState, error) {
+// restoreStateFromDisk returns a mapstate containing the latest
+// snapshot, a flag set to true when the state format has the
+// current version and an error
+func restoreStateFromDisk() (*mapstate.MapState, bool, error) {
 	cfg, _, _, _, consensusCfg, _, _, _, _ := makeConfigs()
 
 	err := cfg.LoadJSONFromFile(configPath)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
 	r, snapExists, err := raft.LastStateRaw(consensusCfg)
 	if !snapExists {
 		err = errNoSnapshot
 	}
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
 	stateFromSnap := mapstate.NewMapState()
-	err = stateFromSnap.Migrate(r)
+	// duplicate reader to both check version and migrate
+	var buf bytes.Buffer
+	r2 := io.TeeReader(r, &buf)
+	raw, err := ioutil.ReadAll(r2)
+	if err != nil {
+		return nil, false, err
+	}
+	err = stateFromSnap.Unmarshal(raw)
```

Comment:
Something is not right here I think. If the state is completely different, `Unmarshal` might just fail because it doesn't know how to do it. That's why we did not unmarshal before and directly call `Migrate`. So it may make migrations fail.

Even if unmarshal does not fail, it might just not replace the `Version` field, which will be set to `mapstate.Version`, and give a false positive when doing `GetVersion() == mapstate.Version` below.

Looking around I see that we have this problem in the `mapstate.Migrate()` function too.

The straightfoward fix is that we make `first-byte-indicates-version` a requirement for all state marshaled-formats, and that we figure out the versions by just reading that first byte, and not by unmarshaling unknown formats on top of the current state types. A state helper (thinking `GetRawVersion(r io.Reader) int`) might help there. We have to remember to avoid loading full states into memory unless we need to (i.e. if we are going to fail or exist, because a version is not the expected, we shouldn't need to do `ReadAll` on the raw state, but just read the first byte.

Reply:
@hsanjuan
> The straightfoward fix is that we make first-byte-indicates-version a requirement for all state marshaled-formats, and that we figure out the versions by just reading that first byte, and not by unmarshaling unknown formats on top of the current state types.

My understanding is that this is what we are doing now.  I haven't been crystal clear about the requirement that first byte (and after a certain point varint as in multiformats) indicates version so thanks for bringing this up.  I should add this more clearly to docs.  As you can see from within Unmarshal however right now we are not unmarshaling unknown formats onto the state type without checking version, if the version doesn't match unmarshal won't decode. 

I agree with you that it is not ideal to read all the bytes before passing to Unmarshal because sometimes the first byte will suffice.  I am good with changing Unmarshal to take in a reader, consume the first byte to check for version, potentially quit if out of date, and continue reading otherwise.  Note that this is going to require some changes (easy I think) to libp2p-raft Restore, and DecodeSnapshot.

Am I understanding all of your concerns correctly?  Did I miss anything?  Are you ok with the suggested fix?  

Thanks!

Reply:
I totally missed that unmarshal checked the version byte before actually unmarshaling because it seemed we were doing that exact check afterwards. It's super weird that it doesn't return an error when it hasn't Unmarshaled anything because of a version mismatch. I think that's what confused me.

Let me have another look tomorrow with a fresh brain.